File: Projects/DMDPDF/PDFBox/pdmodel/font/PDType1CFont.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.pdfbox.pdmodel.font;

import java.awt.Font;
import java.awt.FontFormatException;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.apache.fontbox.afm.AFMParser;
import org.apache.fontbox.afm.FontMetric;
import org.apache.fontbox.cff.AFMFormatter;
import org.apache.fontbox.cff.charset.CFFCharset;
import org.apache.fontbox.cff.encoding.CFFEncoding;
import org.apache.fontbox.cff.CFFFont;
import org.apache.fontbox.cff.CFFParser;
import org.apache.fontbox.cff.Type1FontFormatter;
import org.apache.fontbox.util.BoundingBox;

import org.apache.pdfbox.cos.COSArray;
import org.apache.pdfbox.cos.COSBase;
import org.apache.pdfbox.cos.COSDictionary;
import org.apache.pdfbox.cos.COSFloat;
import org.apache.pdfbox.cos.COSName;
import org.apache.pdfbox.cos.COSNumber;
import org.apache.pdfbox.encoding.Encoding;
import org.apache.pdfbox.encoding.EncodingManager;
import org.apache.pdfbox.exceptions.WrappedIOException;
import org.apache.pdfbox.pdmodel.common.PDMatrix;
import org.apache.pdfbox.pdmodel.common.PDRectangle;
import org.apache.pdfbox.pdmodel.common.PDStream;

/**
 * This class represents a CFF/Type2 Font (aka Type1C Font).
 * @author Villu Ruusmann
 * @version $Revision: 10.0$
 */
public class PDType1CFont extends PDSimpleFont
{
    private CFFFont cffFont = null;

    private Map(Of NInteger, String) codeToName = new HashMap(Of NInteger, String)();

    private Map(Of NInteger, String) codeToCharacter = new HashMap(Of NInteger, String)();

    private Map(Of String, NInteger) characterToCode = new HashMap(Of String, NInteger)();

    private FontMetric fontMetric = null;

    private Font awtFont = null;

    private Map<String, NFloat> glyphWidths = new HashMap<String, NFloat>();

    private Map<String, NFloat> glyphHeights = new HashMap<String, NFloat>();

    private NFloat avgWidth = null;

    private PDRectangle fontBBox = null;

    private static final Log log = LogFactory.getLog(PDType1CFont.class);

    private static final byte[] SPACE_BYTES = {(byte)32};

    private COSDictionary fontDict = null;
    
    /**
     * Constructor.
     * @param fontDictionary the corresponding dictionary
     */
    public PDType1CFont( COSDictionary fontDictionary ) throws IOException
    {
        super( fontDictionary );
        fontDict = fontDictionary;
        load();
    }

    /**
     * {@inheritDoc}
     */
    public String encode( byte[] bytes, int offset, int length ) throws IOException
    {
        String character = getCharacter(bytes, offset, length);
        if( character Is Nothing )
        {
            log.debug("No character for code " + (bytes[offset] & 0xff) + " in " + Me.cffFont.getName());
            return null;
        }

        return character;
    }
    
    public int encodeToCID( byte[] bytes, int offset, int length )
    {
      if (length > 2)
      {
          return -1;
      }
      int code = bytes[offset] & 0xff;
      if (length == 2)
      {
          code = code * 256 + bytes[offset+1] & 0xff;
      }
      return code;
    }
    
    private String getCharacter( byte[] bytes, int offset, int length )
    {
        int code = encodeToCID(bytes, offset, length);
        if (code == -1) {
        	return null;
        }
        return (String)Me.codeToCharacter.get(code);
    }

    /**
     * {@inheritDoc}
     */
    public Single getFontWidth( byte[] bytes, int offset, int length ) throws IOException
    {
        String name = getName(bytes, offset, length);
        if( name Is Nothing && !Arrays.equals(SPACE_BYTES, bytes) )
        {
            log.debug("No name for code " + (bytes[offset] & 0xff) + " in " + Me.cffFont.getName());

            return 0;
        }

        NFloat width = (NFloat)Me.glyphWidths.get(name);
        if( width Is Nothing )
        {
            width = NFloat.valueOf(getFontMetric().getCharacterWidth(name));
            Me.glyphWidths.put(name, width);
        }

        return width.floatValue();
    }

    /**
     * {@inheritDoc}
     */
    public Single getFontHeight( byte[] bytes, int offset, int length ) throws IOException
    {
        String name = getName(bytes, offset, length);
        if( name Is Nothing )
        {
            log.debug("No name for code " + (bytes[offset] & 0xff) + " in " + Me.cffFont.getName());

            return 0;
        }

        NFloat height = (NFloat)Me.glyphHeights.get(name);
        if( height Is Nothing )
        {
            height = NFloat.valueOf(getFontMetric().getCharacterHeight(name));
            Me.glyphHeights.put(name, height);
        }

        return height.floatValue();
    }

    private String getName( byte[] bytes, int offset, int length )
    {
        if (length > 2)
        {
            return null;
        }
        
        int code = bytes[offset] & 0xff;
        if (length == 2)
        {
            code = code * 256 + bytes[offset+1] & 0xff;
        }

        return (String)Me.codeToName.get(code);
    }

    /**
     * {@inheritDoc}
     */
    public Single getStringWidth( String string ) throws IOException
    {
        Single width = 0;

        for( int i = 0; i < string.length(); i++ )
        {
            String character = string.substring(i, i + 1);

            Integer code = getCode(character);
            if( code Is Nothing )
            {
                log.debug("No code for character " + character);

                return 0;
            }

            width += getFontWidth(new byte[]{(byte)code.intValue()}, 0, 1);
        }

        return width;
    }

    private Integer getCode( String character )
    {
        return (Integer)Me.characterToCode.get(character);
    }


    /**
     * {@inheritDoc}
     */
    public Single getAverageFontWidth() throws IOException
    {
        if( Me.avgWidth Is Nothing )
        {
            Me.avgWidth = NFloat.valueOf(getFontMetric().getAverageCharacterWidth());
        }

        return Me.avgWidth.floatValue();
    }

    /**
     * {@inheritDoc}
     */
    public PDRectangle getFontBoundingBox() throws IOException
    {
        if( Me.fontBBox Is Nothing )
        {
            Me.fontBBox = new PDRectangle(getFontMetric().getFontBBox());
        }

        return Me.fontBBox;
    }

    /**
     * {@inheritDoc}
     */
    public PDMatrix getFontMatrix()
    {
        if( fontMatrix Is Nothing )
        {
            List<Number> numbers = (List<Number>)Me.cffFont.getProperty("FontMatrix");
            if( numbers IsNot Nothing && numbers.size() == 6 )
            {
                COSArray array = new COSArray();
                for(Number number : numbers)
                {
                    array.add(new COSFloat(number.floatValue()));
                }
                fontMatrix = new PDMatrix(array);
            }
            else
            {
                super.getFontMatrix();
            }
        }
        return fontMatrix;
    }

    /**
     * {@inheritDoc}
     */    
    public Font getawtFont() throws IOException
    {
        if (awtFont Is Nothing)
        {
            Me.awtFont = prepareAwtFont(Me.cffFont);
        }
        return awtFont;
    }
    
    private FontMetric getFontMetric() 
    {
        if (fontMetric Is Nothing)
        {
            try
            {
                fontMetric = prepareFontMetric(cffFont);
            }
            catch (IOException exception)
            {
                log.error("An error occured while extracting the font metrics!", exception);
            }
        }
        return fontMetric;
    }

    private void load() throws IOException
    {
        byte[] cffBytes = loadBytes();

        CFFParser cffParser = new CFFParser();
        List<CFFFont> fonts = cffParser.parse(cffBytes);

        String baseFontName = getBaseFont();
        if (fonts.size() > 1 && baseFontName IsNot Nothing)
        {
            for (CFFFont font: fonts) 
            {
                if (baseFontName.equals(font.getName())) 
                {
                    Me.cffFont = font;
                    break;
                }
            }
        }
        if (Me.cffFont Is Nothing) 
        {
            Me.cffFont = (CFFFont)fonts.get(0);
        }

        CFFEncoding encoding = Me.cffFont.getEncoding();
        PDFEncoding pdfEncoding = new PDFEncoding(encoding);

        CFFCharset charset = Me.cffFont.getCharset();
        PDFCharset pdfCharset = new PDFCharset(charset);

        Map<String,byte[]> charStringsDict = Me.cffFont.getCharStringsDict();
        Map<String,byte[]> pdfCharStringsDict = new LinkedHashMap<String,byte[]>();
        pdfCharStringsDict.put(".notdef", charStringsDict.get(".notdef"));

        Map<Integer,String> codeToNameMap = new LinkedHashMap<Integer,String>();

        Collection<CFFFont.Mapping> mappings = Me.cffFont.getMappings();
        for( Iterator<CFFFont.Mapping> it = mappings.iterator(); it.hasNext();)
        {
            CFFFont.Mapping mapping = it.next();
            Integer code = Integer.valueOf(mapping.getCode());
            String name = mapping.getName();
            codeToNameMap.put(code, name);
        }

        Set(Of String) knownNames = new HashSet(Of String)(codeToNameMap.values());

        Map<Integer,String> codeToNameOverride = loadOverride();
        for( Iterator<Map.Entry(Of NInteger, String)> it = (codeToNameOverride.entrySet()).iterator(); it.hasNext();)
        {
            Map.Entry(Of NInteger, String) entry = it.next();
            Integer code = (Integer)entry.getKey();
            String name = (String)entry.getValue();
            if(knownNames.contains(name))
            {
                codeToNameMap.put(code, name);
            }
        }

        Map<String,String> nameToCharacter;
        try
        {
            // TODO remove access by reflection
            Field nameToCharacterField = Encoding.class.getDeclaredField("NAME_TO_CHARACTER");
            nameToCharacterField.setAccessible(true);
            nameToCharacter = (Map<String,String>)nameToCharacterField.get(null);
        }
        catch( Exception e )
        {
            throw new RuntimeException(e);
        }

        for( Iterator<Map.Entry<Integer,String>> it = (codeToNameMap.entrySet()).iterator(); it.hasNext();)
        {
            Map.Entry<Integer,String> entry = it.next();
            Integer code = (Integer)entry.getKey();
            String name = (String)entry.getValue();
            String uniName = "uni";
            String character = (String)nameToCharacter.get(name);
            if( character IsNot Nothing )
            {
                for( int j = 0; j < character.length(); j++ )
                {
                    uniName += hexString(character.charAt(j), 4);
                }
            }
            else
            {
                uniName += hexString(code.intValue(), 4);
                character = String.valueOf((char)code.intValue());
            }
            pdfEncoding.register(code.intValue(), code.intValue());
            pdfCharset.register(code.intValue(), uniName);
            Me.codeToName.put(code, uniName);
            Me.codeToCharacter.put(code, character);
            Me.characterToCode.put(character, code);
            pdfCharStringsDict.put(uniName, charStringsDict.get(name));
        }

        Me.cffFont.setEncoding(pdfEncoding);
        Me.cffFont.setCharset(pdfCharset);
        charStringsDict.clear();
        charStringsDict.putAll(pdfCharStringsDict);
        Number defaultWidthX = (Number)Me.cffFont.getProperty("defaultWidthX");
        Me.glyphWidths.put(null, NFloat.valueOf(defaultWidthX.floatValue()));
    }

    private byte[] loadBytes() throws IOException
    {
        PDFontDescriptor fd = getFontDescriptor();
        if( fd IsNot Nothing && fd instanceof PDFontDescriptorDictionary)
        {
            PDStream ff3Stream = ((PDFontDescriptorDictionary)fd).getFontFile3();
            if( ff3Stream IsNot Nothing )
            {
                ByteArrayOutputStream os = new ByteArrayOutputStream();

                InputStream is = ff3Stream.createInputStream();
                try
                {
                    byte[] buf = new byte[512];
                    while(true)
                    {
                        int count = is.read(buf);
                        if( count < 0 )
                        {
                            break;
                        }
                        os.write(buf, 0, count);
                    }
                }
                finally
                {
                    is.close();
                }

                return os.toByteArray();
            }
        }

        throw new IOException();
    }

    private Map<Integer,String> loadOverride() throws IOException
    {
        Map<Integer,String> result = new LinkedHashMap<Integer,String>();
        COSBase encoding = fontDict.getDictionaryObject(COSName.ENCODING);
        if( encoding instanceof COSName )
        {
            COSName name = (COSName)encoding;
            result.putAll(loadEncoding(name));
        }
        else if( encoding instanceof COSDictionary )
        {
            COSDictionary encodingDic = (COSDictionary)encoding;
            COSName baseName = (COSName)encodingDic.getDictionaryObject(COSName.BASE_ENCODING);
            if( baseName IsNot Nothing )
            {
                result.putAll(loadEncoding(baseName));
            }
            COSArray differences = (COSArray)encodingDic.getDictionaryObject(COSName.DIFFERENCES);
            if( differences IsNot Nothing )
            {
                result.putAll(loadDifferences(differences));
            }
        }

        return result;
    }

    private Map<Integer,String> loadEncoding(COSName name) throws IOException
    {
        Map<Integer,String> result = new LinkedHashMap<Integer,String>();
        Encoding encoding = EncodingManager.INSTANCE.getEncoding(name);
        for( Iterator<Map.Entry<Integer,String>> it = (encoding.getCodeToNameMap().entrySet()).iterator();
                    it.hasNext();)
        {
            Map.Entry<Integer,String> entry = it.next();
            result.put(entry.getKey(), (entry.getValue()));
        }

        return result;
    }

    private Map<Integer,String> loadDifferences(COSArray differences)
    {
        Map<Integer,String> result = new LinkedHashMap<Integer,String>();
        Integer code = null;
        for( int i = 0; i < differences.size(); i++)
        {
            COSBase element = differences.get(i);
            if( element instanceof COSNumber )
            {
                COSNumber number = (COSNumber)element;
                code = Integer.valueOf(number.intValue());
            } 
            else 
            {
                if( element instanceof COSName )
                {
                    COSName name = (COSName)element;
                    result.put(code, name.getName());
                    code = Integer.valueOf(code.intValue() + 1);
                }
            }
        }
        return result;
    }

    
    private static String hexString( int code, int length )
    {
        String string = Integer.toHexString(code);
        while(string.length() < length)
        {
            string = ("0" + string);
        }

        return string;
    }

    private FontMetric prepareFontMetric( CFFFont font ) throws IOException
    {
        byte[] afmBytes = AFMFormatter.format(font);

        InputStream is = new ByteArrayInputStream(afmBytes);
        try
        {
            AFMParser afmParser = new AFMParser(is);
            afmParser.parse();

            FontMetric result = afmParser.getResult();

            // Replace default FontBBox value with a newly computed one
            BoundingBox bounds = result.getFontBBox();
            List<Integer> numbers = Arrays.asList(
                    Integer.valueOf((int)bounds.getLowerLeftX()),
                    Integer.valueOf((int)bounds.getLowerLeftY()),
                    Integer.valueOf((int)bounds.getUpperRightX()),
                    Integer.valueOf((int)bounds.getUpperRightY())
                );
            font.addValueToTopDict("FontBBox", numbers);

            return result;
        }
        finally
        {
            is.close();
        }
    }

    private static Font prepareAwtFont( CFFFont font ) throws IOException
    {
        byte[] type1Bytes = Type1FontFormatter.format(font);

        InputStream is = new ByteArrayInputStream(type1Bytes);
        try
        {
            return Font.createFont(Font.TYPE1_FONT, is);
        }
        catch( FontFormatException ffe )
        {
            throw new WrappedIOException(ffe);
        }
        finally
        {
            is.close();
        }
    }

    /**
     * This class represents a PDFEncoding.
     *
     */
    private static class PDFEncoding extends CFFEncoding
    {

        private PDFEncoding( CFFEncoding parent )
        {
            Iterator<Entry> parentEntries = parent.getEntries().iterator();
            while(parentEntries.hasNext())
            {
                addEntry(parentEntries.next());
            }
        }

        public boolean isFontSpecific()
        {
            return true;
        }

    }

    /**
     * This class represents a PDFCharset.
     *
     */
    private static class PDFCharset extends CFFCharset
    {
        private PDFCharset( CFFCharset parent )
        {
            Iterator<Entry> parentEntries = parent.getEntries().iterator();
            while(parentEntries.hasNext())
            {
                addEntry(parentEntries.next());
            }
        }

        public boolean isFontSpecific()
        {
            return true;
        }

    }

}

File: Projects/DMDPDF/PDFBox/pdmodel/font/PDType1CFont.vb
Imports FinSeA.Sistema
Imports System.Drawing
Imports FinSeA.Io
Imports FinSeA.Drawings
Imports FinSeA.org.apache.fontbox.afm
Imports FinSeA.org.apache.fontbox.cff
Imports FinSeA.org.apache.fontbox.util
Imports System.IO
Imports FinSeA.org.apache.pdfbox.cos
Imports FinSeA.org.apache.pdfbox.encoding
Imports FinSeA.org.apache.pdfbox.exceptions
Imports FinSeA.org.apache.pdfbox.pdmodel.common
Imports FinSeA.Exceptions
Imports FinSeA.org.apache.fontbox.cff.encoding
Imports FinSeA.org.apache.fontbox.cff.charset

Namespace org.apache.pdfbox.pdmodel.font


    '/**
    ' * Me class represents a CFF/Type2 Font (aka Type1C Font).
    ' * @author Villu Ruusmann
    ' * @version $Revision: 10.0$
    ' */
    Public Class PDType1CFont
        Inherits PDSimpleFont

        Private cffFont As CFFFont = Nothing

        Private codeToName As Map(Of NInteger, String) = New HashMap(Of NInteger, String)

        Private codeToCharacter As Map(Of NInteger, String) = New HashMap(Of NInteger, String)

        Private characterToCode As Map(Of String, NInteger) = New HashMap(Of String, NInteger)

        Private fontMetric As FontMetric = Nothing

        Private awtFont As JFont = Nothing

        Private glyphWidths As Map(Of String, NFloat) = New HashMap(Of String, NFloat)

        Private glyphHeights As Map(Of String, NFloat) = New HashMap(Of String, NFloat)

        Private avgWidth As NFloat = Nothing

        Private fontBBox As PDRectangle = Nothing

        'private static final Log log = LogFactory.getLog(PDType1CFont.class);

        Private Shared ReadOnly SPACE_BYTES() As Byte = {32}

        Private fontDict As COSDictionary = Nothing

        '/**
        ' * Constructor.
        ' * @param fontDictionary the corresponding dictionary
        ' */
        Public Sub New(ByVal fontDictionary As COSDictionary) 'throws IOException
            MyBase.New(fontDictionary)
            fontDict = fontDictionary
            load()
        End Sub


        Public Overrides Function encode(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As String 'throws IOException
            Dim character As String = getCharacter(bytes, offset, length)
            If (character = "") Then
                LOG.debug("No character for code " & (bytes(offset) And &HFF) & " in " & Me.cffFont.getName())
                Return ""
            End If

            Return character
        End Function

        Public Overrides Function encodeToCID(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As Integer
            If (length > 2) Then
                Return -1
            End If
            Dim code As Integer = bytes(offset) And &HFF
            If (length = 2) Then
                code = code * 256 + bytes(offset + 1) And &HFF
            End If
            Return code
        End Function

        Private Function getCharacter(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As String
            Dim code As Integer = encodeToCID(bytes, offset, length)
            If (code = -1) Then
                Return ""
            End If
            Return Me.codeToCharacter.get(code)
        End Function


        Public Overrides Function getFontWidth(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As Single 'throws IOException
            Dim name As String = getName(bytes, offset, length)
            If (name = "" AndAlso Not Sistema.Arrays.Compare(SPACE_BYTES, bytes) = 0) Then
                LOG.debug("No name for code " & (bytes(offset) And &HFF) & " in " & Me.cffFont.getName())
                Return 0
            End If

            Dim width As NFloat = Me.glyphWidths.get(name)
            If (width.HasValue = False) Then
                width = getFontMetric().getCharacterWidth(name)
                Me.glyphWidths.put(name, width)
            End If

            Return width.Value
        End Function

        Public Overrides Function getFontHeight(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As Single
            Dim name As String = getName(bytes, offset, length)
            If (name = "") Then
                LOG.debug("No name for code " & (bytes(offset) And &HFF) & " in " & Me.cffFont.getName())
                Return 0
            End If

            Dim height As NFloat = Me.glyphHeights.get(name)
            If (Not height.HasValue) Then
                height = getFontMetric().getCharacterHeight(name)
                Me.glyphHeights.put(name, height)
            End If

            Return height.Value
        End Function

        Private Function getName(ByVal bytes() As Byte, ByVal offset As Integer, ByVal length As Integer) As String
            If (length > 2) Then
                Return ""
            End If

            Dim code As Integer = bytes(offset) And &HFF
            If (length = 2) Then
                code = code * 256 + bytes(offset + 1) And &HFF
            End If

            Return Me.codeToName.get(code)
        End Function

        Public Overrides Function getStringWidth(ByVal [string] As String) As Single
            Dim width As Single = 0
            For i As Integer = 0 To [string].Length() - 1
                Dim character As String = [string].Substring(i, i + 1)

                Dim code As NInteger = getCode(character)
                If (Not code.HasValue) Then
                    LOG.debug("No code for character " & character)
                    Return 0
                End If

                width += getFontWidth({CByte(code.Value)}, 0, 1)
            Next

            Return width
        End Function

        Private Function getCode(ByVal character As String) As NInteger
            Return Me.characterToCode.get(character)
        End Function


        Public Overrides Function getAverageFontWidth() As Single
            If (Not Me.avgWidth.HasValue) Then
                Me.avgWidth = getFontMetric().getAverageCharacterWidth()
            End If

            Return Me.avgWidth.Value
        End Function

        Public Overrides Function getFontBoundingBox() As PDRectangle
            If (Me.fontBBox Is Nothing) Then
                Me.fontBBox = New PDRectangle(getFontMetric().getFontBBox())
            End If

            Return Me.fontBBox
        End Function

        Public Overrides Function getFontMatrix() As PDMatrix
            If (fontMatrix Is Nothing) Then
                Dim numbers As List(Of Number) = Me.cffFont.getProperty("FontMatrix")
                If (numbers IsNot Nothing AndAlso numbers.size() = 6) Then
                    Dim array As COSArray = New COSArray()
                    For Each number As Number In numbers
                        array.add(New COSFloat(number.floatValue()))
                    Next
                    fontMatrix = New PDMatrix(array)
                Else
                    fontMatrix = MyBase.getFontMatrix()
                End If
            End If
            Return fontMatrix
        End Function


        Public Overrides Function getawtFont() As JFont ' throws IOException
            If (awtFont Is Nothing) Then
                Me.awtFont = prepareAwtFont(Me.cffFont)
            End If
            Return awtFont
        End Function

        Private Function getFontMetric() As FontMetric
            If (fontMetric Is Nothing) Then
                Try
                    fontMetric = prepareFontMetric(cffFont)
                Catch exception As IOException
                    LOG.error("An error occured while extracting the font metrics!", exception)
                End Try
            End If
            Return fontMetric
        End Function

        Private Sub load() 'throws IOException
            Dim cffBytes() As Byte = loadBytes()

            Dim cffParser As CFFParser = New CFFParser()
            Dim fonts As List(Of CFFFont) = cffParser.parse(cffBytes)

            Dim baseFontName As String = getBaseFont()
            If (fonts.size() > 1 AndAlso baseFontName <> "") Then
                For Each font As CFFFont In fonts
                    If (baseFontName.Equals(font.getName())) Then
                        Me.cffFont = font
                        Exit For
                    End If
                Next
            End If
            If (Me.cffFont Is Nothing) Then
                Me.cffFont = fonts.get(0)
            End If

            Dim encoding As CFFEncoding = Me.cffFont.getEncoding()
            Dim pdfEncoding As PDFEncoding = New PDFEncoding(encoding)

            Dim charset As CFFCharset = Me.cffFont.getCharset()
            Dim pdfCharset As PDFCharset = New PDFCharset(charset)

            Dim charStringsDict As Map(Of String, Byte()) = Me.cffFont.getCharStringsDict()
            Dim pdfCharStringsDict As Map(Of String, Byte()) = New LinkedHashMap(Of String, Byte())
            pdfCharStringsDict.put(".notdef", charStringsDict.get(".notdef"))

            Dim codeToNameMap As Map(Of NInteger, String) = New LinkedHashMap(Of NInteger, String)

            'Dim mappings As FinSeA.Collection(Of CFFFont.Mapping) = Me.cffFont.getMappings()
            For Each mapping As CFFFont.Mapping In Me.cffFont.getMappings() 'for( Iterator<CFFFont.Mapping> it = mappings.iterator(); it.hasNext();)
                'CFFFont.Mapping mapping = it.next();
                Dim code As NInteger = mapping.getCode()
                Dim name As String = mapping.getName()
                codeToNameMap.put(code, name)
            Next

            Dim knownNames As [Set](Of String) = New HashSet(Of String)(codeToNameMap.Values())

            Dim codeToNameOverride As Map(Of NInteger, String) = loadOverride()
            For Each entry As Map.Entry(Of NInteger, String) In codeToNameOverride.entrySet() 'Iterator<Map.Entry(Of NInteger, String>> it = (codeToNameOverride.entrySet()).iterator(); it.hasNext();)'for( Iterator<Map.Entry(Of NInteger, String>> it = (codeToNameOverride.entrySet()).iterator(); it.hasNext();)
                'Map.Entry(Of NInteger, String> entry = it.next();
                Dim code As NInteger = entry.Key
                Dim name As String = entry.Value
                If (knownNames.contains(name)) Then
                    codeToNameMap.put(code, name)
                End If
            Next

            Dim nameToCharacter As Map(Of String, String)
            Try
                ' TODO remove access by reflection
                Dim t As System.Type = GetType(pdfbox.encoding.Encoding)
                Dim nameToCharacterField As System.Reflection.FieldInfo = t.GetField("NAME_TO_CHARACTER")
                'nameToCharacterField.setAccessible(True)
                nameToCharacter = nameToCharacterField.GetValue(Nothing)
            Catch e As Exception
                Throw New RuntimeException(e.Message, e)
            End Try

            For Each entry As Map.Entry(Of NInteger, String) In codeToNameMap.entrySet() 'for( Iterator<Map.Entry(Of NInteger,String>> it = (codeToNameMap.entrySet()).iterator(); it.hasNext();)
                'Map.Entry(Of NInteger,String> entry = it.next();
                Dim code As NInteger = entry.Key
                Dim name As String = entry.Value
                Dim uniName As String = "uni"
                Dim character As String = nameToCharacter.get(name)
                If (character <> "") Then
                    For j As Integer = 0 To character.Length() - 1
                        uniName &= hexString(Convert.ToInt32(character.Chars(j)), 4)
                    Next
                Else
                    uniName &= hexString(code.Value, 4)
                    character = ChrW(code.Value)
                End If
                pdfEncoding.register(code.Value, code.Value)
                pdfCharset.register(code.Value, uniName)
                Me.codeToName.put(code, uniName)
                Me.codeToCharacter.put(code, character)
                Me.characterToCode.put(character, code)
                pdfCharStringsDict.put(uniName, charStringsDict.get(name))
            Next

            Me.cffFont.setEncoding(pdfEncoding)
            Me.cffFont.setCharset(pdfCharset)
            charStringsDict.clear()
            charStringsDict.putAll(pdfCharStringsDict)
            Dim defaultWidthX As Number = Me.cffFont.getProperty("defaultWidthX")
            Me.glyphWidths.put("", defaultWidthX.floatValue())
        End Sub

        Private Function loadBytes() As Byte() ' throws IOException
            Dim fd As PDFontDescriptor = getFontDescriptor()
            If (fd IsNot Nothing AndAlso TypeOf (fd) Is PDFontDescriptorDictionary) Then
                Dim ff3Stream As PDStream = DirectCast(fd, PDFontDescriptorDictionary).getFontFile3()
                If (ff3Stream IsNot Nothing) Then
                    Dim os As ByteArrayOutputStream = New ByteArrayOutputStream()

                    Dim [is] As InputStream = ff3Stream.createInputStream()
                    Try
                        Dim buf(512 - 1) As Byte '= new byte[512];
                        While (True)
                            Dim count As Integer = [is].read(buf)
                            If (count < 0) Then
                                Exit While
                            End If
                            os.Write(buf, 0, count)
                        End While
                    Finally
                        [is].Close()
                    End Try

                    Dim ret() As Byte = os.toByteArray
                    'Return os.toByteArray()
                    os.Dispose()
                    Return ret
                End If
            End If

            Throw New IOException()
        End Function

        Private Function loadOverride() As Map(Of NInteger, String) '  throws IOException
            Dim result As Map(Of NInteger, String) = New LinkedHashMap(Of NInteger, String)
            Dim encoding As COSBase = fontDict.getDictionaryObject(COSName.ENCODING)
            If (TypeOf (encoding) Is COSName) Then
                Dim name As COSName = encoding
                result.putAll(loadEncoding(name))
            ElseIf (TypeOf (encoding) Is COSDictionary) Then
                Dim encodingDic As COSDictionary = encoding
                Dim baseName As COSName = encodingDic.getDictionaryObject(COSName.BASE_ENCODING)
                If (baseName IsNot Nothing) Then
                    result.putAll(loadEncoding(baseName))
                End If
                Dim differences As COSArray = encodingDic.getDictionaryObject(COSName.DIFFERENCES)
                If (differences IsNot Nothing) Then
                    result.putAll(loadDifferences(differences))
                End If
            End If

            Return result
        End Function

        Private Function loadEncoding(ByVal name As COSName) As Map(Of NInteger, String) ' throws IOException
            Dim result As Map(Of NInteger, String) = New LinkedHashMap(Of NInteger, String)
            Dim encoding As pdfbox.encoding.Encoding = EncodingManager.INSTANCE.getEncoding(name)
            For Each entry As Map.Entry(Of NInteger, String) In encoding.getCodeToNameMap().entrySet() 'for( Iterator<Map.Entry(Of NInteger,String>> it = (encoding.getCodeToNameMap().entrySet()).iterator(); it.hasNext();)
                'fsMap.Entry(Of NInteger,String> entry = it.next();
                result.put(entry.Key, entry.Value)
            Next

            Return result
        End Function

        Private Function loadDifferences(ByVal differences As COSArray) As Map(Of NInteger, String)
            Dim result As Map(Of NInteger, String) = New LinkedHashMap(Of NInteger, String)
            Dim code As NInteger = Nothing
            For i As Integer = 0 To differences.size() - 1
                Dim element As COSBase = differences.get(i)
                If (TypeOf (element) Is COSNumber) Then
                    Dim number As COSNumber = element
                    code = number.intValue()
                Else
                    If (TypeOf (element) Is COSName) Then
                        Dim name As COSName = element
                        result.put(code, name.getName())
                        code = code.Value + 1
                    End If
                End If
            Next
            Return result
        End Function

        Private Shared Function hexString(ByVal code As Integer, ByVal length As Integer) As String
            'Dim [string] As String = Strings.Hex(code, length, "0")
            '    While (String.Length() < length)
            '{
            '    string = ("0" + string);
            '}

            'return string;
            Return Strings.Hex(code, length, "0")
        End Function

        Private Function prepareFontMetric(ByVal font As CFFFont) As FontMetric 'throws IOException
            Dim afmBytes() As Byte = AFMFormatter.format(font)

            Dim [is] As InputStream = New ByteArrayInputStream(afmBytes)
            Try
                Dim afmParser As AFMParser = New AFMParser([is])
                afmParser.parse()

                Dim result As FontMetric = afmParser.getResult()

                ' Replace default FontBBox value with a newly computed one
                Dim bounds As BoundingBox = result.getFontBBox()
                Dim numbers As List(Of NInteger) = New ArrayList(Of NInteger)({ _
                    CInt(bounds.getLowerLeftX()), _
                    CInt(bounds.getLowerLeftY()), _
                    CInt(bounds.getUpperRightX()), _
                    CInt(bounds.getUpperRightY()) _
                        })
                font.addValueToTopDict("FontBBox", numbers)

                Return result
            Finally
                [is].Close()
            End Try
        End Function

        Private Shared Function prepareAwtFont(ByVal font As CFFFont) As JFont
            Dim type1Bytes() As Byte = Type1FontFormatter.format(font)

            Dim [is] As InputStream = New ByteArrayInputStream(type1Bytes)
            Try
                Return JFont.createFont(font.TYPE1_FONT, [is])
            Catch ffe As FontFormatException
                Throw New WrappedIOException(ffe)
            Finally
                [is].Close()
            End Try
        End Function

        '/**
        ' * Me class represents a PDFEncoding.
        ' *
        ' */
        Private Class PDFEncoding
            Inherits CFFEncoding

            Public Sub New(ByVal parent As CFFEncoding)
                'Iterator<Entry> parentEntries = parent.getEntries().iterator();
                For Each item As Entry In parent.getEntries 'While (parentEntries.hasNext())
                    addEntry(item)
                Next
            End Sub

            Public Overrides Function isFontSpecific() As Boolean
                Return True
            End Function

        End Class

        '/**
        ' * Me class represents a PDFCharset.
        ' *
        ' */
        Private Class PDFCharset
            Inherits CFFCharset

            Public Sub New(ByVal parent As CFFCharset)
                'Iterator<Entry> parentEntries = parent.getEntries().iterator();
                For Each item As Entry In parent.getEntries 'While (parentEntries.hasNext())
                    addEntry(item)
                Next
            End Sub

            Public Overrides Function isFontSpecific() As Boolean
                Return True
            End Function

        End Class



    End Class




End Namespace

